### PR TITLE
API-key-rotation-Update-API-key-changes

### PR DESCRIPTION
**Pull request by GunaKKIBM**

### CODE REVIEW

File: provider/local/provider.go [hunk 1]
```diff
@@ -36,6 +36,9 @@ type Provider interface {
 
 	// Returns a configured ContextCredentialsFactory for this provider
 	ContextCredentialsFactory(datacenter *string) (ContextCredentialsFactory, error)
+
+	// UpdateAPIKey updates api key in the provider
+	UpdateAPIKey(*vpcconfig.VPCBlockConfig, *zap.Logger) error
```
Comment (sameshai) [inline provider/local/provider.go:41]: In case of file are planning to introduce one more variant of UpdateAPIKey as it will be using different config.

Comment (GunaKKIBM) [reply provider/local/provider.go:41]: I have an idea for the same. Will raise another PR for this

File: provider/iam/token_exchange.go [hunk 2]
```diff
@@ -172,6 +172,12 @@ func (tes *tokenExchangeService) newTokenExchangeRequest(logger *zap.Logger) *to
 	}
 }
 
+// UpdateAPIKey ...
+func (tes *tokenExchangeService) UpdateAPIKey(apiKey string, logger *zap.Logger) error {
+	// This method need not be implemented
```
Comment (sameshai) [inline provider/iam/token_exchange.go:177]: When we take the non-service account route, this token_exchange.go is used. Can we implement the method ?

File: provider/local/provider.go (hunk 1 above)
Comment (sameshai) [reply provider/local/provider.go:41]: Ok we can take this as improvment

File: lib/provider/error_codes.go [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Package provider ...
+package provider
+
+const (
+	// APIKeyNotFound ...
+	APIKeyNotFound = "E06f9"
```
Comment (sameshai) [inline lib/provider/error_codes.go:22]: Can we see if we need to add the below error code 
```
{"level":"error","ts":"2021-11-08T18:22:29.053Z","caller":"iam/token_exchange_iks.go:169","msg":"IAM token exchange request failed with message","RequestID":"dc423289-5961-494f-8f1f-b889ade04d60","StatusCode":400,"API IncidentID":"d4a453df-94ae-ac1f-b722-8cfbed6a4a79","Error":{"code":"E5fae","description":"Error message from IAM: '400 Bad Request. Transaction-Id: aWFtaWQtNi4xMC0xMjA4OC0yYmIxMDU4LTdkNzQ3Y2NkODctMnRoNDI-e0b3e0f553f044e8b7158c2a2fc20a67 Details: {\"errorCode\":\"BXNIM0410E\",\"errorMessage\":\"Provided user not found or active\",\"context\":{\"requestId\":\"aWFtaWQtNi4xMC0xMjA4OC0yYmIxMDU4LTdkNzQ3Y2NkODctMnRoNDI-e0b3e0f553f044e8b7158c2a2fc20a67\",\"requestType\":\"incoming.Identity_Token\",\"userAgent\":\"Go-http-client/1.1\",\"url\":\"https://iam.bluemix.net\",\"instanceId\":\"iamid-6.10-12088-2bb1058-7d747ccd87-2th42\",\"threadId\":\"1e7f74\",\"host\":\"iamid-6.10-12088-2bb1058-7d747ccd87-2th42\",\"startTime\":\"08.11.2021 18:22:28:327 UTC\",\"endTime\":\"08.11.2021 18:22:29:031 UTC\",\"elapsedTime\":\"704\",\"locale\":\"en_US\",\"clusterName\":\"iam-id-prod-us-south-dal12\"}}'.","type":"Authentication","incidentID":"d4a453df-94ae-ac1f-b722-8cfbed6a4a79"}}
{"level":"error","ts":"2021-11-08T18:22:29.053Z","caller":"auth/ims_token.go:73","msg":"Unable to retrieve IAM access token from IAM API key","RequestID":"dc423289-5961-494f-8f1f-b889ade04d60","error":"IAM token exchange request failed: Error message from IAM: '400 Bad Request. Transaction-Id: aWFtaWQtNi4xMC0xMjA4OC0yYmIxMDU4LTdkNzQ3Y2NkODctMnRoNDI-e0b3e0f553f044e8b7158c2a2fc20a67 Details: {\"errorCode\":\"BXNIM0410E\",\"errorMessage\":\"Provided user not found or active\",\"context\":{\"requestId\":\"aWFtaWQtNi4xMC0xMjA4OC0yYmIxMDU4LTdkNzQ3Y2NkODctMnRoNDI-e0b3e0f553f044e8b7158c2a2fc20a67\",\"requestType\":\"incoming.Identity_Token\",\"userAgent\":\"Go-http-client/1.1\",\"url\":\"https://iam.bluemix.net\",\"instanceId\":\"iamid-6.10-12088-2bb1058-7d747ccd87-2th42\",\"threadId\":\"1e7f74\",\"host\":\"iamid-6.10-12088-2bb1058-7d747ccd87-2th42\",\"startTime\":\"08.11.2021 18:22:28:327 UTC\",\"endTime\":\"08.11.2021 18:22:29:031 UTC\",\"elapsedTime\":\"704\",\"locale\":\"en_US\",\"clusterName\":\"iam-id-prod-us-south-dal12\"}}'."}
{"level":"error","ts":"2021-11-08T18:22:29.053Z","caller":"provider/provider.go:77","msg":"Error occurred while generating IAM token for VPC","RequestID":"dc423289-5961-494f-8f1f-b889ade04d60","error":"IAM token exchange request failed: Error message from IAM: '400 Bad Request. Transaction-Id: aWFtaWQtNi4xMC0xMjA4OC0yYmIxMDU4LTdkNzQ3Y2NkODctMnRoNDI-e0b3e0f553f044e8b7158c2a2fc20a67 Details: {\"errorCode\":\"BXNIM0410E\",\"errorMessage\":\"Provided user not found or active\",\"context\":{\"requestId\":\"aWFtaWQtNi4xMC0xMjA4OC0yYmIxMDU4LTdkNzQ3Y2NkODctMnRoNDI-e0b3e0f553f044e8b7158c2a2fc20a67\",\"requestType\":\"incoming.Identity_Token\",\"userAgent\":\"Go-http-client/1.1\",\"url\":\"https://iam.bluemix.net\",\"instanceId\":\"iamid-6.10-12088-2bb1058-7d747ccd87-2th42\",\"threadId\":\"1e7f74\",\"host\":\"iamid-6.10-12088-2bb1058-7d747ccd87-2th42\",\"startTime\":\"08.11.2021 18:22:28:327 UTC\",\"endTime\":\"08.11.2021 18:22:29:031 UTC\",\"elapsedTime\":\"704\",\"locale\":\"en_US\",\"clusterName\":\"iam-id-prod-us-south-dal12\"}}'."}
{"level":
```

File: provider/iam/token_exchange.go (hunk 2 above)
Comment (GunaKKIBM) [reply provider/iam/token_exchange.go:177]: Will consider in the upcoming PRs.

**Review by prankulmahajan — APPROVED** [review]
lgtm

**Comment by GunaKKIBM** [comment]
Test results:

1. Creating pvc by deleting api key, initially, the pvc is in pending state, kubernetes retries to create volume. Meanwhile, if the api key is reset, volume gets created. It won't be necessary to restart the pod. 
```
kubectl create -f pvc.yaml                                   
persistentvolumeclaim/csi-vpc-block-5iops-pvc-api-key-rotation40 created

gunakk@Gunas-MBP OPENSHIFT_VPC_CLUSTER_TESTING % kubectl describe pvc csi-vpc-block-5iops-pvc-api-key-rotation40 
Name:          csi-vpc-block-5iops-pvc-api-key-rotation40
Namespace:     default
StorageClass:  ibmc-vpc-block-5iops-tier
Status:        Bound
Volume:        pvc-d7da1591-4aec-474a-b5f2-a406fe4b815c
Labels:        <none>
Annotations:   pv.kubernetes.io/bind-completed: yes
               pv.kubernetes.io/bound-by-controller: yes
               volume.beta.kubernetes.io/storage-provisioner: vpc.block.csi.ibm.io
Finalizers:    [kubernetes.io/pvc-protection]
Capacity:      10Gi
Access Modes:  RWO
VolumeMode:    Filesystem
Used By:       <none>
Events:
  Type     Reason                 Age                   From                                                                                      Message
  ----     ------                 ----                  ----                                                                                      -------
  Warning  ProvisioningFailed     3m29s                 vpc.block.csi.ibm.io_ibm-vpc-block-csi-controller-0_1d3a9aa8-1bc4-4e29-90cc-1e07d9edfef1  failed to provision volume with StorageClass "ibmc-vpc-block-5iops-tier": rpc error: code = Internal desc = {RequestID: 9d2c7752-c1e4-491a-a61f-fbb45a30b7d7 , Code: InternalError, Description: Internal error occurred, BackendError: API key is not found, reset the api key. If reset, retry after few minutes, Action: Please check 'BackendError' tag for more details}
  Warning  ProvisioningFailed     2m29s                 vpc.block.csi.ibm.io_ibm-vpc-block-csi-controller-0_1d3a9aa8-1bc4-4e29-90cc-1e07d9edfef1  failed to provision volume with StorageClass "ibmc-vpc-block-5iops-tier": rpc error: code = Internal desc = {RequestID: 077e111a-e15c-46d4-a54e-0abc0cf919d7 , Code: InternalError, Description: Internal error occurred, BackendError: API key is not found, reset the api key. If reset, retry after few minutes, Action: Please check 'BackendError' tag for more details}
  Normal   Provisioning           88s (x4 over 4m29s)   vpc.block.csi.ibm.io_ibm-vpc-block-csi-controller-0_1d3a9aa8-1bc4-4e29-90cc-1e07d9edfef1  External provisioner is provisioning volume for claim "default/csi-vpc-block-5iops-pvc-api-key-rotation40"
  Warning  ProvisioningFailed     88s                   vpc.block.csi.ibm.io_ibm-vpc-block-csi-controller-0_1d3a9aa8-1bc4-4e29-90cc-1e07d9edfef1  failed to provision volume with StorageClass "ibmc-vpc-block-5iops-tier": rpc error: code = Internal desc = {RequestID: b1c7a84a-04ec-4f74-be16-47ef86202d18 , Code: InternalError, Description: Internal error occurred, BackendError: API key is not found, reset the api key. If reset, retry after few minutes, Action: Please check 'BackendError' tag for more details}
  Normal   ExternalProvisioning   20s (x18 over 4m29s)  persistentvolume-controller                                                               waiting for a volume to be created, either by external provisioner "vpc.block.csi.ibm.io" or manually created by system administrator
  Normal   ProvisioningSucceeded  11s                   vpc.block.csi.ibm.io_ibm-vpc-block-csi-controller-0_1d3a9aa8-1bc4-4e29-90cc-1e07d9edfef1  Successfully provisioned volume pvc-d7da1591-4aec-474a-b5f2-a406fe4b815c

 % kubectl create -f validPOD2.yaml
deployment.apps/api-key-rotation-pod40 created

 % kubectl get pods
NAME                                      READY   STATUS             RESTARTS   AGE
api-key-rotation-pod40-6fff7bb57-vmnxn    1/1     Running            0          3m26s
```

**Pull request merged**